include/scope.hpp

```diff
@@ -27,57 +27,49 @@ enum class ScopeKind : std::uint8_t {
 /// @brief A symbol table associated with a scope kind.
 struct Scope {
   ScopeKind kind;
-  std::unique_ptr<SymbolTable> table;
+  std::unique_ptr<SymbolTable> symbol_table;
+  std::unique_ptr<DeclTypeTable> type_table;
 
-  Scope(ScopeKind kind, std::unique_ptr<SymbolTable> table)
-      : kind{kind}, table{std::move(table)} {}
+  Scope(ScopeKind kind, std::unique_ptr<SymbolTable> symbol_table,
+        std::unique_ptr<DeclTypeTable> type_table)
+      : kind{kind},
+        symbol_table{std::move(symbol_table)},
+        type_table{std::move(type_table)} {}
 };
 
 /// @brief Manages scopes and symbol tables.
 class ScopeStack {
  public:
   /// @brief Pushes a new scope of the kind.
-  /// @throws `ScopesOfDifferentKindIsNotMergeableError` if the previous scope
-  /// had set to merge with the next (this) scope, which is of a different kind.
-  void PushScope(ScopeKind kind) {
-    if (should_merge_with_next_scope_) {
-      if (scopes_.back().kind != kind) {
-        throw ScopesOfDifferentKindIsNotMergeableError{""};
-      }
-      should_merge_with_next_scope_ = false;
-      // No need to push a new scope.
-      return;
-    }
-    scopes_.emplace_back(kind, std::make_unique<SymbolTable>());
-  }
-
-  /// @throws `NotInScopeError`
-  void PopScope() {
-    ThrowIfNotInScope_();
-    scopes_.pop_back();
-  }
+  void PushScope(ScopeKind kind);
+  /// @brief Pops the top scope of the stack.
+  void PopScope();
 
   /// @brief Adds an entry to the top-most scope of the kind.
   /// @throws `NotInScopeError`
   /// @throws `NotInSuchKindOfScopeError`
-  std::shared_ptr<SymbolEntry> Add(std::unique_ptr<SymbolEntry> entry,
-                                   ScopeKind kind) {
+  template <typename Table, typename EntryType>
+  std::shared_ptr<EntryType> AddEntry(std::unique_ptr<EntryType> entry,
+                                      ScopeKind kind,
+                                      std::unique_ptr<Table> Scope::*table) {
     ThrowIfNotInScope_();
     for (auto it = scopes_.rbegin(); it != scopes_.rend(); ++it) {
       if (it->kind == kind) {
-        return it->table->Add(std::move(entry));
+        return ((*it).*table)->Add(std::move(entry));
       }
     }
     throw NotInSuchKindOfScopeError{""};
   }
 
   /// @brief Looks up the `id` from through all scopes.
   /// @throws `NotInScopeError`
-  std::shared_ptr<SymbolEntry> LookUp(const std::string& id) const {
+  template <typename Table, typename EntryType>
```

Comment:
Can simplify to `Entry`. It's obvious that this is a type.
```suggestion
  template <typename Table, typename Entry>
```